user-interface/src/main/java/life/qbic/datamanager/views/projects/project/measurements/MeasurementMain.java

```diff
@@ -418,33 +403,62 @@ private void handleDeletionResults(Result<Void, MeasurementDeletionException> re
     result.onValue(v -> measurementDetailsComponent.refreshGrids());
   }
 
-  private void executeInUi(Command command) {
-    getUI().ifPresent(ui -> ui.access(command));
-  }
-
   private void downloadProteomicsMetadata(List<String> selectedMeasurementIds) {
+    UI ui = getUI().orElse(null);
+    if (ui == null) {
+      return; // UI has been detached
+    }
     ProjectId projectId = context.projectId().orElseThrow();
     var inProgressToast = messageFactory.pendingTaskToast("measurement.preparing-download",
         MessageSourceNotificationFactory.EMPTY_PARAMETERS, getLocale());
+
+    if (ui.getSession() != null && ui.getSession().hasLock()) {
+      inProgressToast.open();
+    } else {
+      ui.access(inProgressToast::open);
+    }
+
     asyncService.measurementUpdatePxP(projectId.value(), selectedMeasurementIds, OPEN_XML)
-        .doOnSubscribe(ignore -> executeInUi(inProgressToast::open))
-        .doOnSuccess(this::triggerDownload)
-        .doFinally(it -> executeInUi(inProgressToast::close))
-        .subscribe();
+        .subscribe(result -> ui.access(() -> {
+          inProgressToast.close();
+          ui.push();
+          triggerDownload(result, ui);
+        }), error -> {
+          ui.access(inProgressToast::close);
+          log.error(error.getMessage(), error);
+        }, () -> ui.access(inProgressToast::close));
   }
 
   private void downloadNGSMetadata(List<String> selectedMeasurementIds) {
+    UI ui = getUI().orElse(null);
+    if (ui == null) {
+      return; // UI has been detached
+    }
     ProjectId projectId = context.projectId().orElseThrow();
     var inProgressToast = messageFactory.pendingTaskToast("measurement.preparing-download",
         MessageSourceNotificationFactory.EMPTY_PARAMETERS, getLocale());
-    asyncService.measurementUpdateNGS(projectId.value(), selectedMeasurementIds, OPEN_XML)
-        .doOnSubscribe(ignore -> executeInUi(inProgressToast::open))
-        .doOnSuccess(this::triggerDownload)
-        .doFinally(it -> executeInUi(inProgressToast::close))
-        .subscribe();
-  }
 
-  private void triggerDownload(DigitalObject digitalObject) {
+    if (ui.getSession() != null && ui.getSession().hasLock()) {
+      inProgressToast.open();
+    } else {
+      ui.access(inProgressToast::open);
+    }
+
+    asyncService.measurementUpdateNGS(projectId.value(), selectedMeasurementIds, OPEN_XML)
+        .subscribe(result -> ui.access(() -> {
+          inProgressToast.close();
+          ui.push();
```

Comment:
why do you enforce a ui push before the download is triggered?
For reactor flux or mono, vaadin recommends using UI#accessLater https://vaadin.com/docs/latest/building-apps/deep-dives/presentation-layer/server-push/reactive